File: packages/provide-uterm-go/bridge/hijackable.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package bridge

import (
	"context"
	"log/slog"
	"sync"
	"time"
)

// Port of provide.uterm.bridge.base.HijackableMixin.
//
// Go has no mixin/multiple-inheritance, so the Python mixin becomes a struct a
// worker embeds (or holds) to gain checkpoint hijackability. A worker calls
// AwaitIfHijacked at checkpoints in its automation loop; a hub/manager calls
// SetHijacked to pause/resume; a dashboard calls RequestStep to release one
// (or a few) checkpoint passes while still paused.
//
// Deviation from Python: AwaitIfHijacked takes a context so a blocked
// checkpoint can be cancelled (the Python coroutine relies on task
// cancellation). The watchdog callback (OnStuck) is a plain func and any panic
// it raises is recovered and logged, mirroring the Python except-and-log.

// gate is a broadcast latch modelled on asyncio.Event. A "set" gate lets
// waiters through immediately; a "clear" gate blocks them until set. It starts
// set (matching HijackableMixin: not hijacked by default).
type gate struct {
	mu sync.Mutex
	ch chan struct{}
}

func newGate() *gate {
	ch := make(chan struct{})
	close(ch) // start set → no blocking
	return &gate{ch: ch}
}

// set releases every current and future waiter until the next clear.
func (g *gate) set() {
	g.mu.Lock()
	defer g.mu.Unlock()
	select {
	case <-g.ch: // already set (closed)
	default:
		close(g.ch)
	}
}

// clear makes waiters block until the next set.
func (g *gate) clear() {
	g.mu.Lock()
	defer g.mu.Unlock()
	select {
	case <-g.ch: // currently set (closed) → install a fresh open channel
		g.ch = make(chan struct{})
	default: // already clear
	}
}

// wait blocks until the gate is set or ctx is cancelled.
func (g *gate) wait(ctx context.Context) error {
	g.mu.Lock()
	ch := g.ch
	g.mu.Unlock()
	select {
	case <-ch:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

// WatchdogOptions configure StartWatchdog. Zero values select the Python
// defaults (StuckTimeout 120s, CheckInterval 5s, no callback).
type WatchdogOptions struct {
	// StuckTimeout is the idle duration before OnStuck fires.
	StuckTimeout time.Duration
	// CheckInterval is how often the watchdog polls (floored at 500ms).
	CheckInterval time.Duration
	// OnStuck is invoked when the worker stops making progress. A panic it
	// raises is recovered and logged.
	OnStuck func()
}

// Hijackable provides pause/resume/step/watchdog primitives for a worker.
type Hijackable struct {
	mu           sync.Mutex
	hijacked     bool
	stepTokens   int
	lastProgress time.Time
	gate         *gate

	watchdogCancel context.CancelFunc
	watchdogDone   chan struct{}

	logger *slog.Logger
}

// NewHijackable builds a Hijackable. logger may be nil (a no-op logger is
// substituted). The gate starts set, so a worker is not hijacked by default.
func NewHijackable(logger *slog.Logger) *Hijackable {
	if logger == nil {
		logger = slog.New(slog.DiscardHandler)
	}
	return &Hijackable{
		gate:         newGate(),
		lastProgress: time.Now(),
		logger:       logger,
	}
}

// AwaitIfHijacked blocks the automation while a human is hijacking this worker.
//
// Call it at every checkpoint in the automation loop. It returns immediately
// when not hijacked or when a step token is available (consuming one),
// otherwise it blocks until SetHijacked(false) or ctx cancellation. Port of
// await_if_hijacked.
func (h *Hijackable) AwaitIfHijacked(ctx context.Context) error {
	h.mu.Lock()
	if !h.hijacked {
		h.mu.Unlock()
		return nil
	}
	if h.stepTokens > 0 {
		h.stepTokens--
		h.mu.Unlock()
		return nil
	}
	h.mu.Unlock()
	return h.gate.wait(ctx)
}

// SetHijacked pauses (true) or resumes (false) automation. It is idempotent.
// Port of set_hijacked.
func (h *Hijackable) SetHijacked(enabled bool) {
	h.mu.Lock()
	defer h.mu.Unlock()
	if enabled == h.hijacked {
		return
	}
	h.hijacked = enabled
	if enabled {
		h.stepTokens = 0
		h.gate.clear()
	} else {
		h.gate.set()
	}
}

// RequestStep allows automation to pass checkpoints hijack gates while still
// hijacked. Two tokens (the Python default) let one full loop iteration run;
// tokens are capped at 100 and a non-positive argument is treated as zero. It
// is a no-op when not hijacked. Port of request_step.
//
// Like Python, RequestStep does not wake a checkpoint that is already blocked
// on the gate: tokens are consumed at the next checkpoint reached while the
// automation is running.
func (h *Hijackable) RequestStep(checkpoints int) {
	h.mu.Lock()
	defer h.mu.Unlock()
	if !h.hijacked {
		return
	}
	if checkpoints < 0 {
		checkpoints = 0
	}
	h.stepTokens = min(h.stepTokens+checkpoints, 100)
}

// IsHijacked reports whether the worker is currently paused.
func (h *Hijackable) IsHijacked() bool {
	h.mu.Lock()
	defer h.mu.Unlock()
	return h.hijacked
}

// StepTokens returns the number of buffered step tokens (test/introspection).
func (h *Hijackable) StepTokens() int {
	h.mu.Lock()
	defer h.mu.Unlock()
	return h.stepTokens
}

// NoteProgress resets the watchdog timer, signalling meaningful work. Port of
// note_progress.
func (h *Hijackable) NoteProgress() {
	h.mu.Lock()
	h.lastProgress = time.Now()
	h.mu.Unlock()
}

// StartWatchdog starts a background goroutine that fires opts.OnStuck if no
// NoteProgress call has been seen for opts.StuckTimeout. While hijacked, the
// timer is suppressed. Calling it again while a watchdog runs is a no-op. Port
// of start_watchdog.
func (h *Hijackable) StartWatchdog(opts WatchdogOptions) {
	if opts.StuckTimeout <= 0 {
		opts.StuckTimeout = 120 * time.Second
	}
	if opts.CheckInterval <= 0 {
		opts.CheckInterval = 5 * time.Second
	}
	if opts.CheckInterval < 500*time.Millisecond {
		opts.CheckInterval = 500 * time.Millisecond
	}

	h.mu.Lock()
	if h.watchdogCancel != nil {
		h.mu.Unlock()
		return
	}
	ctx, cancel := context.WithCancel(context.Background())
	done := make(chan struct{})
	h.watchdogCancel = cancel
	h.watchdogDone = done
	h.mu.Unlock()

	go h.watchdogLoop(ctx, done, opts)
}

func (h *Hijackable) watchdogLoop(ctx context.Context, done chan struct{}, opts WatchdogOptions) {
	defer close(done)
	ticker := time.NewTicker(opts.CheckInterval)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
		}
		h.mu.Lock()
		hijacked := h.hijacked
		last := h.lastProgress
		h.mu.Unlock()
		if hijacked {
			h.NoteProgress()
			continue
		}
		if time.Since(last) < opts.StuckTimeout {
			continue
		}
		if opts.OnStuck != nil {
			h.callOnStuck(opts.OnStuck)
		}
		// Reset so we do not spam if reconnect is slow.
		h.NoteProgress()
	}
}

// callOnStuck invokes the callback, recovering and logging any panic (mirroring
// the Python except-and-log around the on_stuck callback).
func (h *Hijackable) callOnStuck(fn func()) {
	defer func() {
		if r := recover(); r != nil {
			h.logger.Warn("watchdog on_stuck callback raised", "panic", r)
		}
	}()
	fn()
}

// StopWatchdog cancels the watchdog goroutine and waits for it to exit. It is
// idempotent. Port of stop_watchdog.
func (h *Hijackable) StopWatchdog() {
	h.mu.Lock()
	cancel := h.watchdogCancel
	done := h.watchdogDone
	h.watchdogCancel = nil
	h.watchdogDone = nil
	h.mu.Unlock()
	if cancel != nil {
		cancel()
	}
	if done != nil {
		<-done
	}
}

// CleanupHijack releases the hijack and stops the watchdog. Call from a
// worker's cleanup/shutdown path. Port of cleanup_hijack.
func (h *Hijackable) CleanupHijack() {
	h.SetHijacked(false)
	h.StopWatchdog()
}

File: packages/provide-uterm-ts/src/bridge/hijackable.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * Making an automated worker pausable by a human operator.
 *
 * Port of the Python module `provide.uterm.bridge.base`.
 *
 * A worker calls {@link Hijackable.awaitIfHijacked} at each checkpoint in its
 * loop. Normally that returns at once; while an operator holds the session it
 * blocks, and the dashboard can let the loop through one iteration at a time.
 *
 * The reference is a mixin because Python has cooperative multiple
 * inheritance. Here it is a plain object a worker holds, which is the same
 * capability without the diamond.
 */

/**
 * Most step tokens that can be banked at once.
 *
 * Unbounded accumulation from a client hammering the step button would let
 * the loop run away the moment the hijack is released.
 */
export const STEP_TOKEN_CAP = 100;

/**
 * Checkpoints one press of Step buys.
 *
 * Two — plan and act — so a single press advances the loop exactly one
 * iteration rather than half of one.
 */
export const STEP_TOKENS_PER_ITERATION = 2;

/** Default seconds without progress before the watchdog fires. */
const DEFAULT_STUCK_TIMEOUT_S = 120;

/** Default seconds between watchdog checks. */
const DEFAULT_CHECK_INTERVAL_S = 5;

/** Shortest check interval; below this the loop spins rather than watches. */
const MIN_CHECK_INTERVAL_S = 0.5;

/** Options for {@link Hijackable}. */
export interface HijackableOptions {
  /** Monotonic clock in seconds. */
  now?: () => number;
  /** How the watchdog waits between checks. */
  sleep?: (seconds: number) => Promise<void>;
}

/** Options for {@link Hijackable.startWatchdog}. */
export interface WatchdogOptions {
  /** Seconds without progress before firing. */
  stuckTimeoutS?: number;
  /** Seconds between checks, floored at {@link MIN_CHECK_INTERVAL_S}. */
  checkIntervalS?: number;
  /** Called when the worker looks stuck. Typically drops the session. */
  onStuck?: () => Promise<void>;
}

/** Monotonic seconds, from a clock that cannot jump backwards. */
function monotonicNow(): number {
  return performance.now() / 1000;
}

/** Makes an automated worker pausable by a human operator. */
export class Hijackable {
  readonly #now: () => number;
  readonly #sleep: (seconds: number) => Promise<void>;
  #hijacked = false;
  #stepTokens = 0;
  #lastProgress: number;
  /** Resolvers for checkpoints currently blocked. */
  #waiters: Array<() => void> = [];
  #watchdog: { stop: () => void; done: Promise<void> } | undefined;

  constructor(options: HijackableOptions = {}) {
    this.#now = options.now ?? monotonicNow;
    this.#sleep = options.sleep ?? ((seconds) => new Promise((resolve) => setTimeout(resolve, seconds * 1000)));
    this.#lastProgress = this.#now();
  }

  /** Whether an operator currently holds the worker. */
  get hijacked(): boolean {
    return this.#hijacked;
  }

  /** Checkpoints the worker may still pass while held. */
  get stepTokens(): number {
    return this.#stepTokens;
  }

  /**
   * A checkpoint in the worker's loop.
   *
   * Returns at once when nobody holds the worker, spends a step token when
   * one is banked, and otherwise blocks until the hijack ends — which is the
   * whole point: the automation stops where the operator put it.
   */
  async awaitIfHijacked(): Promise<void> {
    if (!this.#hijacked) {
      return;
    }
    if (this.#stepTokens > 0) {
      this.#stepTokens -= 1;
      return;
    }
    await new Promise<void>((resolve) => {
      this.#waiters.push(resolve);
    });
  }

  /**
   * Pause or resume the worker. Idempotent.
   *
   * Taking the worker discards any tokens granted for a previous hold, so an
   * operator pausing again does not find the loop already walking.
   */
  async setHijacked(enabled: boolean): Promise<void> {
    if (enabled === this.#hijacked) {
      return;
    }
    this.#hijacked = enabled;
    if (enabled) {
      this.#stepTokens = 0;
      return;
    }
    const waiters = this.#waiters;
    this.#waiters = [];
    for (const resolve of waiters) {
      resolve();
    }
  }

  /**
   * Let the worker through `checkpoints` more gates while still held.
   *
   * A no-op when nobody holds it: banking tokens beforehand would let a later
   * hijack be walked straight through. A negative request adds nothing rather
   * than debiting credit already granted, which would otherwise let a client
   * drive the count below zero and unblock the gate for good.
   */
  async requestStep(checkpoints: number = STEP_TOKENS_PER_ITERATION): Promise<void> {
    if (!this.#hijacked) {
      return;
    }
    const requested = Math.max(0, Math.trunc(checkpoints));
    this.#stepTokens = Math.min(this.#stepTokens + requested, STEP_TOKEN_CAP);
  }

  /** Note that the worker did something, resetting the watchdog. */
  noteProgress(): void {
    this.#lastProgress = this.#now();
  }

  /**
   * Watch for the worker going quiet, and call `onStuck` when it does.
   *
   * Suppressed while the worker is held: a paused worker is not a stuck one,
   * and firing then would drop the very session the operator is driving.
   *
   * The timer resets after firing, so a slow reconnect is not spammed with
   * repeat callbacks. A second start is ignored while one is already running.
   */
  startWatchdog(options: WatchdogOptions = {}): void {
    if (this.#watchdog !== undefined) {
      return;
    }
    const stuckTimeoutS = options.stuckTimeoutS ?? DEFAULT_STUCK_TIMEOUT_S;
    const intervalS = Math.max(MIN_CHECK_INTERVAL_S, options.checkIntervalS ?? DEFAULT_CHECK_INTERVAL_S);
    let running = true;

    const done = (async () => {
      while (running) {
        await this.#sleep(intervalS);
        if (!running) {
          return;
        }
        if (this.#hijacked) {
          // Held, not stalled — keep the clock fresh so releasing the hold
          // does not immediately look like a stall.
          this.noteProgress();
          continue;
        }
        if (this.#now() - this.#lastProgress < stuckTimeoutS) {
          continue;
        }
        if (options.onStuck !== undefined) {
          try {
            await options.onStuck();
          } catch {
            // The callback is recovery code; if it fails there is nothing
            // better to do than keep watching.
          }
        }
        this.noteProgress();
      }
    })();

    this.#watchdog = {
      stop: () => {
        running = false;
      },
      done,
    };
  }

  /** Stop the watchdog. Idempotent. */
  async stopWatchdog(): Promise<void> {
    const watchdog = this.#watchdog;
    if (watchdog === undefined) {
      return;
    }
    this.#watchdog = undefined;
    watchdog.stop();
  }

  /**
   * Release the worker and stop watching it.
   *
   * Called from the worker's shutdown, so neither a paused loop nor a live
   * timer is left behind.
   */
  async cleanupHijack(): Promise<void> {
    await this.setHijacked(false);
    await this.stopWatchdog();
  }
}
